### PR TITLE
Log UUID inclusion, fixes #23

### LINKED ISSUE #23: Add (optional) UUID to log name
This proposal is to add the ability for the user to request a UUID in the log name and optionally turn off the hostname e.g. `logstash_2018-01-01T01:01:00.part1.{{UUID}}.log`. This has the following benefits:

1. A new log won't clobber the old one after Logstash is restarted on a host with the same name. See  @neuromantik33 description in #20 for more info.
2. Users could turn off using the hostname in their file names if:
  * They don't want an information leak: @joshdvir opened #9 for this
  * They needed to be able to use a single glob to match their logs to import into BigQuery @arslanm had to modify the plugin in #19 to do this

Thoughts @neuromantik33, @joshdvir and @arslanm?

### PR DESCRIPTION
**Pull request by josephlewis42**
Adds new flags to turn on/off the hostname in the log file name and whether or not to include a UUID in the file. This will allow users to match files with a single wildcard (to aid them in importing to BigQuery or other systems) and ensures files won't clobber one another if multiple instances of Logstash are running or an instance cycles up and down quickly.

### CODE REVIEW

**Comment by josephlewis42** [comment]
@rosbo would you take a look when you get a chance? It's a slight refactor and two new pieces of functionality plus tests (yay!).

It should pave the way towards solving some of our longest standing issues around locking. It'll open up the way towards fixing #2 (from 2015) and #19 which will in turn open up #22 fixing #5. It'll also open up the possibility of changing upload file names to be whatever the user wants, including directories to fix #24.

I'm also hopeful that the pieces will be independent enough they could be pulled into bigquery to offer a "free mode" once again as an alternative to the streaming uploads.

File: spec/outputs/gcs/path_factory_spec.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      expected = 'prefix_2018-01-01.log'
+      expected = File.join('path/to/directory', expected)
+
+      actual = pf.current_path(vars)
+
+      expect(actual).to eq(expected)
+    end
+
+    it 'loads a path immediately' do
+      pf = LogStash::Outputs::Gcs::PathFactory.new('', '', false, '', false, false, false)
```
Comment (rosbo) [inline spec/outputs/gcs/path_factory_spec.rb]: It is hard to follow what is going on here. Consider using a builder pattern to initialize the `PathFactory`?
https://medium.com/kkempin/builder-design-pattern-in-ruby-dfa2d557ff1b

**Comment by josephlewis42** [comment]
I really like how easy it is to generate those builders in Ruby, good call!

The builds are failing right now for Logstash master and 6.X, they seem to have broken something upstream because it's also happening for Pub/Sub and bigquery.

**Comment by jsvd** [comment]
@josephlewis42 we sorted out the issues that were causing global failures in plugin tests, we can merge this now

**Comment by josephlewis42** [comment]
@jsvd merged, built and tagged as v3.1.0, would you mind kicking off a release to rubygems? (unless the tag automatically does that)

**Comment by josephlewis42** [comment]
@jsvd That would be awesome! `jlewisiii@google.com` should work, same `josephlewis42` username as here.

**Comment by jsvd** [comment]
@josephlewis42 done: https://rubygems.org/gems/logstash-output-google_cloud_storage
